tests/dragonfly/conftest.py

```diff
@@ -334,38 +337,80 @@ def with_ca_tls_client_args(with_tls_client_args, with_tls_ca_cert_args):
+        # Global tracking of all failed tests/logs
+        failed_list = open("/tmp/failed_list.txt", "a")
         files = last_log_file.readlines()
         logging.error(f"Test failed {report.nodeid} with logs: ")
         for file in files:
-            # copy to failed folder
+            failed_list.write(file)
             file = file.rstrip("\n")
             logging.error(f"🪵🪵🪵🪵🪵🪵 {file} 🪵🪵🪵🪵🪵🪵")
-            shutil.copy(file, failed_path)
-
-
-def pytest_exception_interact(node, call, report):
-    if report.failed:
-        copy_failed_logs_and_clean_tmp_folder(report)
-
 
-@pytest.fixture(autouse=True)
+        # Clean it
+        last_log_file = open("/tmp/last_test_log_files.txt", "w").close()
+        global TEST_FAILED
+        TEST_FAILED = True
+
+
+# Double consider any change to this function because the order
+# of SetUp and TearDown might cause improper copy of the logs
+# because the instance has not yet stopped. For example, df_factory
+# has session scope so it won't be called unless the session has ended.
+# On the other hand, df_local_factory has function scope, and therefore
+# it is executed strictly before this (so the instance is stoped and we can
+# safely copy the logs without loosing some of them).
+# autouse means that this is run for all tests in the module.
+# scope="session" allows to create the file only once and clean it
+# at the end respecting the evaluation order mentioned above
+@pytest.fixture(autouse=True, scope="session")
 def run_before_and_after_test():
-    # Setup: logic before any of the test starts
-    # Empty the log on each run
-    last_log_file = open("/tmp/last_test_log_files.txt", "w")
-    last_log_file.close()
+    logging.info("Session start for run_before_and_after_test")
+    # Setup: at the start of the session
+    last_log_file = open("/tmp/last_test_log_files.txt", "w").close()
 
     yield  # this is where the testing happens
 
-    # Teardown
+    global TEST_FAILED
+    # Teardown at the end of the session
+    logging.info(f"Session end for run_before_and_after_test")
+    if TEST_FAILED:
+        logging.info(f"Copying failed tests to /tmp/failed")
+        copy_failed_logs_and_clean_tmp_folder()
+
+
+# The only issue here is that this also clears the logs for session wide fixtures which
+# could be problematic for tests that failed way later in the session. Example:
+# suppose a session fixture like dragonfly_factory run by 3 individual tests
+# First one instantiates the dragonfly instance and writes its log path in the
+# last_test_log_files.txt. Second test then clears this, and third one fails without any logs.
+# This problem existed before but luckilly so far hasn't come up. There is no really good fix for
+# this as it's hard to track which test used which fixture when it failed, since we only interact
+# with the failed test via the pytest hook pytest_exception_interact.
+@pytest.fixture(autouse=True, scope="function")
```

Comment:
We can see this cascading here: https://github.com/dragonflydb/dragonfly/actions/runs/9681657533/job/26712730878?pr=3226#step:12:905

There are two solutions for this:

1. Keep a global list of session wide fixtures and their log paths. Upload them all when a test case fails but we will have to manually find them which means looking at the CI the port and grepping the big list of uploaded files
2. Track down where the same instance/fixture is used multiple times and either replace a session fixture with a function fixture and see if this increases total time of tests.
3. We can create a per function fixture that polls the session dfly instances and temporary append them to the path. That way the local failure will contain the interesting logs

I am leaning towards 3 because it can solve this nicely. I will play around but on a different PR. It's an issue on it's own that existed before this PR.

Reply:
fixed on the last commit